src/video.js

```diff
@@ -64,3 +69,76 @@ export const checkVideoBidSetup = hook('sync', function(bid, adUnit, videoMediaT
 
   return true;
 }, 'checkVideoBidSetup');
+
+export function registerVastTrackers(moduleType, moduleName, trackerFn) {
+  if (typeof trackerFn == 'function') { vastTrackers.push({'moduleType': moduleType, 'moduleName': moduleName, 'trackerFn': trackerFn}); }
+}
+
+export function insertVastTrackers(trackers, vastXml) {
+  const doc = new DOMParser().parseFromString(vastXml, 'text/xml');
+  const wrappers = doc.querySelectorAll('VAST Ad Wrapper, VAST Ad InLine');
+  try {
+    if (wrappers.length) {
+      wrappers.forEach(wrapper => {
+        if (trackers.get('impressions')) {
+          trackers.get('impressions').forEach(trackingUrl => {
+            const impression = doc.createElement('Impression');
+            impression.appendChild(doc.createCDATASection(trackingUrl));
+            wrapper.appendChild(impression)
+          });
+        }
+      });
+      vastXml = new XMLSerializer().serializeToString(doc);
+    }
+  } catch (error) {
+    logError('an error happened trying to insert trackers in vastXml');
+  }
+  return vastXml;
+}
+
+export function getVastTrackers(bid) {
+  let trackers = [];
+  vastTrackers.filter(
+    ({moduleType, moduleName, trackerFn}) => isActivityAllowed(ACTIVITY_REPORT_ANALYTICS, activityParams(moduleType, moduleName))
+  ).forEach(({trackerFn}) => {
+    let trackersToAdd = trackerFn(bid);
+    trackersToAdd.forEach(trackerToAdd => {
+      if (isValidVastTracker(trackers, trackerToAdd)) { trackers.push(trackerToAdd); }
+    });
+  });
+  const trackersMap = trackersToMap(trackers);
+  return (trackersMap.size ? trackersMap : null);
+};
+
+function isValidVastTracker(trackers, trackerToAdd) {
+  if (!trackerToAdd.hasOwnProperty('event') || !trackerToAdd.hasOwnProperty('url')) { return false; }
+  return true;
+}
+
+// need to add this othersize gulp test for test/spec/video_spec.js and test/spec/videoCache_spec.js duplicates their trackings
+function isTrackingDuplicate(trackersMap, trackingEvent, trackingUrl) {
+  const urls = trackersMap.get(trackingEvent);
+  if (Array.isArray(urls) && urls.includes(trackingUrl)) { return true; }
+  return false;
+}
+
+function trackersToMap(trackers) {
+  let trackersMap = new Map();
+  trackers.forEach(tracker => {
+    if (!trackersMap.get(tracker['event'])) {
+      trackersMap.set(tracker['event'], [tracker['url']])
+    } else if (!isTrackingDuplicate(trackersMap, tracker['event'], tracker['url'])) {
+      trackersMap.get(tracker['event']).push(tracker['url']);
+    }
+  });
+  return trackersMap;
+}
+
+export function addImpUrlToTrackers(bid, trackersMap) {
```

Comment:
this could be replaced by a single line in `getVastTrackers instead`:

```javascript
// ... just before calling trackersToMap
  bid.vastImpUrl && trackers.push({event: 'impression', url: bid.vastImpUrl});
  const trackersMap = trackersToMap(trackers);
// ...

```

Reply:
Adding `bid.vastImpUrl` is not mandatory that's the reason why I've made is through a dedicated function which is call only when required. The current use case is when the video needs to be cached and there is no `bid.vastXml` object.